### PR TITLE
POC: Allow guarding on unbacked symints

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90985
* #92107

This PR demonstrates a simple (perhaps too simple) approach to dealing with guards on unbacked symints. We maintain a list of substitutions of expressions with unbacked symints, and then when we fail to statically evaluate an expression because it contains unbacked symints, we attempt to apply those substitutions to see if they can eliminate the unbacked symints. The test case manually programs the substitutions in the ShapeEnv, but in general we need some more user friendly API for doing so.

I am not sure if I should actually finish preparing this patch, however, because as you can see the user experience for what substitutions you need to add is very, very bad. It turns out, we do a LOT of 0/1 specializing.

Fixes https://github.com/pytorch/pytorch/issues/91990

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by ezyang** [comment]
cc @nunoplopes 

I think the simplification strategy here is too simple. We should do some of Nuno's stuff.

**Comment by ezyang** [comment]
Now updated with nonzero and boolean masking

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -392,6 +392,28 @@ def local_scalar_dense(fake_mode, func, arg):
...
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz < arg.numel()).node.expr, False))
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((arg.numel() * nnz == 0).node.expr, False))
+    # TODO: the substitutions below are unsound
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz == 0).node.expr, False))
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz == 1).node.expr, False))
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz * arg.dim() == 0).node.expr, False))
+    # TODO: these are just hacks for the example
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((nnz < 2).node.expr, False))
+    fake_mode.shape_env.expr_subs[nnz.node.expr].append(((9216 * nnz).node.expr, False))
+    return arg.new_empty((nnz, arg.dim()), dtype=torch.int64)
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:414]: I like, kind of suspect maybe the unbacked symint allocation should happen in meta tensor, not fake tensor. Main problem is getting our grubby hands on the ambient shapenv. Maybe torch._guards is the ticket.

File: test/test_proxy_tensor.py [hunk 2]
```diff
@@ -892,6 +892,38 @@ def forward(self, a_1):
     mul = torch.ops.aten.mul.Tensor(a_1, _local_scalar_dense);  a_1 = _local_scalar_dense = None
     return mul""")
 
+    def test_boolean_index(self):
+        def f(images, handedness):
+            right_hand_mask = handedness == 1
+            images[right_hand_mask] = images[right_hand_mask].flip(-1)
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_proxy_tensor.py:898]: @ngimel and I discussed this and in fact it is not so easy to do so, because you need to materialize the boolean masked `images[right_hand_mask]` to flip it.

[the remark this answers is not in the document]
Comment (lezcano) [reply test/test_proxy_tensor.py:898]: Can we ask Philip to give us the power of sorting?
```python
import torch

def f(img, mask):
    img[mask] = img[mask].flip(-1)
    return img

def g(img, mask):
    assert mask.ndim == 1
    # Move true elements to the front and flip them
    sorted_flip = mask.argsort(stable=True).flip(-1)
    sorted = mask.argsort(descending=True, stable=True).argsort()
    return torch.where(mask, img[sorted_flip[sorted]], img)


img = torch.randn(10)
mask = torch.tensor([True, False, False, False, True, True, False, True, True, False])
print(mask)
print(f(img.clone(), mask.clone()))
print(g(img.clone(), mask.clone()))
```

I don't claim that this is a *great* solution. This is just a proof of concept that it can be done without dynamic shapes.

[the remark this answers is not in the document]
Comment (lezcano) [reply test/test_proxy_tensor.py:898]: Also, note that via this pattern you can implement very many functions on a subset of elements selected by a mask.

[the remark this answers is not in the document]
Comment (ngimel) [reply test/test_proxy_tensor.py:898]: Yeah if you want to return a different output and not inplace modify `images` as above it's theoretically doable (but you'd still need a global sort)

[the remark this answers is not in the document]
Comment (lezcano) [reply test/test_proxy_tensor.py:898]: On a second thought, this may be viable even without sorting.

Note that a stable sort on a boolean mask is clearly O(n).

Then, the two sorts computing `sorted` just compute the inverse permutation, so these two sorts combined can also be computed in one linear pass.

All in all, the function `g` can be implemented in two for loops. The first for loop would implement
```python
sorted1 = mask.argsort(stable=True)
sorted = mask.argsort(descending=True, stable=True).argsort()
```
and the second one would do
```
torch.where(mask, img[sorted1.flip(-1)[sorted]], img)
```

The only issue here would be parallelising the first for loop in triton. In CUDA, implementing that first for loop is not as difficult as implementing a regular parallel sorting algorithm, but it's not trivially parallelisable either, so I don't think this is implementable in triton with the current machinery.
